### PR TITLE
[builtins] Fix divtc3.c etc. compilation on Solaris/SPARC with gcc

### PR DESCRIPTION
**Pull request by rorth**
`compiler-rt/lib/builtins/divtc3.c` and `multc3.c` don't compile on Solaris/sparcv9 with `gcc -m32`:
```
FAILED: projects/compiler-rt/lib/builtins/CMakeFiles/clang_rt.builtins-sparc.dir/divtc3.c.o
[...]
compiler-rt/lib/builtins/divtc3.c: In function ‘__divtc3’:
compiler-rt/lib/builtins/divtc3.c:22:18: error: implicit declaration of function ‘__compiler_rt_logbtf’ [-Wimplicit-function-declaration]
   22 |   fp_t __logbw = __compiler_rt_logbtf(
      |                  ^~~~~~~~~~~~~~~~~~~~
```
and many more.  It turns out that while the definition of `__divtc3` is guarded with `CRT_HAS_F128`, the `__compiler_rt_logbtf` and other declarations use `CRT_HAS_128BIT && CRT_HAS_F128` as guard.  This only shows up with `gcc` since, as documented in Issue #41838, `clang` violates the SPARC psABI in not using 128-bit `long double`, so this code path isn't used.

Fixed by changing the guards to match.

Tested on `sparcv9-sun-solaris2.11`.

### CODE REVIEW

**Review by arichardson — APPROVED** [review]

**Comment by zibi2** [comment]
@rorth Rainer, this change broke the following libcxx'`s lit tests on z/OS:

```
std/numerics/complex.number/complex.transcendentals/pow_complex_complex.pass.cpp
std/numerics/complex.number/complex.member.ops/divide_equal_complex.pass.cpp
std/numerics/complex.number/complex.transcendentals/pow_complex_scalar.pass.cpp
std/numerics/complex.number/complex.ops/scalar_divide_complex.pass.cpp
std/numerics/complex.number/complex.transcendentals/atan.pass.cpp
std/numerics/complex.number/complex.transcendentals/atanh.pass.cpp
std/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp
std/numerics/complex.number/complex.member.ops/times_equal_complex.pass.cpp
std/numerics/complex.number/complex.transcendentals/pow_scalar_complex.pass.cpp
std/numerics/complex.number/cmplx.over/pow.pass.cpp
```

They all failed since after this change `__divtc3` and `__multc3` are not available. It looks like only `CRT_HAS_F128` macro is defined. which was used originally to guard definitions of those functions. 

On z/OS we don't define CRT_HAD_128BIT and according to this [snippet code](https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/builtins/int_types.h#L71C1-L76C7) from `int)types.h` Microsoft as well so they might be affected as well.

Would you be able to revert this change and come up with the proper fix for this?

**Comment by zibi2** [comment]
In a nutshell the issue is when `CRT_HAS_F128` is defined but `CRT_HAS_128BIT` has not.

**Comment by rorth** [comment]
> @rorth Rainer, this change broke the following libcxx'`s lit tests on z/OS:
[...]

Sorry for the failure.

> They all failed since after this change `__divtc3` and `__multc3` are not available. It looks like only `CRT_HAS_F128` macro is defined. which was used originally to guard definitions of those functions.
> 
> On z/OS we don't define CRT_HAD_128BIT and according to this [snippet code](https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/builtins/int_types.h#L71C1-L76C7) from `int)types.h` Microsoft as well so they might be affected as well.

Me knowing nothing about z/OS and not having access to such a system, could you please provide the output of the compilation of `divtc3.c` with `-g3 -dD -save-temps` added?  Going through the maze of macro definitions was bad enough when I had to on SPARC, on a foreign system it's pure guesswork.

> Would you be able to revert this change and come up with the proper fix for this?

Given that the patch has been in for two weeks without complaint, could you give me a day or two to investigate and come up with a fix?  This avoid the constant churn of Revert/Reland/Re-Revert/Re-Reland.  Thanks.

**Comment by rorth** [comment]
> > Me knowing nothing about z/OS and not having access to such a system, could you please provide the output of the compilation of `divtc3.c` with `-g3 -dD -save-temps` added? Going through the maze of macro definitions was bad enough when I had to on SPARC, on a foreign system it's pure guesswork.
> 
> Thank you for responding. The -save-temps`does not work. I'm going to attached output from compiling`divtc3.c`with`--dD -E` added. [divthc3.c.txt](https://github.com/user-attachments/files/16625525/divthc3.c.txt)

`-save-temps` probably did work, but what's confusing is that the `.i` file is stored alongside the `.o`, not in the working directory as one would expect.

There are a few things that confuse me, though: 
- I find `CRT_HAS_128BIT` first defined (from `int_types.h` since `__LP64__` is defined), that immediately undefined.       However, in the source, that's guarded like this:
  ```
  #if defined(_MSC_VER) && !defined(__clang__)
  #undef CRT_HAS_128BIT
  #endif
  ```
  However, in the `.i` file `_MSC_VER` isn't defined while `__clang__` is.
- The `.i` file claims to be from `compiler-rt/lib/builtins/divthc3.c`, which files doesn't exist upstream.  The body of the file isn't anything like `divtc3.c` either.
Whatever the case, when looking at my patch again, I found that it should have guarded the `__divtc3` and `__multc3` definitions like this (relative to current sources):
```
diff --git a/compiler-rt/lib/builtins/divtc3.c b/compiler-rt/lib/builtins/divtc3.c
--- a/compiler-rt/lib/builtins/divtc3.c
+++ b/compiler-rt/lib/builtins/divtc3.c
@@ -13,7 +13,7 @@
 #define QUAD_PRECISION
 #include "fp_lib.h"
 
-#if defined(CRT_HAS_128BIT) && defined(CRT_HAS_F128)
+#if defined(CRT_HAS_F128) && defined(CRT_HAS_TF_MODE)
 
 // Returns: the quotient of (a + ib) / (c + id)
 
diff --git a/compiler-rt/lib/builtins/multc3.c b/compiler-rt/lib/builtins/multc3.c
--- a/compiler-rt/lib/builtins/multc3.c
+++ b/compiler-rt/lib/builtins/multc3.c
@@ -15,7 +15,7 @@
 #include "int_lib.h"
 #include "int_math.h"
 
-#if defined(CRT_HAS_128BIT) && defined(CRT_HAS_F128)
+#if defined(CRT_HAS_F128) && defined(CRT_HAS_TF_MODE)
 
 // Returns: the product of a + ib and c + id
 
```
All the functions use (`__compiler_rt_logbtf` and several more) are guarded by `CRT_HAS_TF_MODE`.  That works on SPARC, but won't for you since it isn't defined in your `.i` file.
Are you maybe using modified sources?
> > Given that the patch has been in for two weeks without complaint, could you give me a day or two to investigate and come up with a fix? This avoid the constant churn of Revert/Reland/Re-Revert/Re-Reland. Thanks.
> 
> Waiting fews days should be fine with us,.

Good, thanks.

**Comment by zibi2** [comment]
> `-save-temps` probably did work, but what's confusing is that the `.i` file is stored alongside the `.o`, not in the working directory as one would expect.

Unfortunately the `.i` is not created due to unrelated errors.

> * I find `CRT_HAS_128BIT` first defined (from `int_types.h` since `__LP64__` is defined), that immediately undefined.       However, in the source, that's guarded like this:
>   ```
>   #if defined(_MSC_VER) && !defined(__clang__)
>   #undef CRT_HAS_128BIT
>   #endif

Our downstream code has modified guard as follows:
```
#if (defined(_MSC_VER) && !defined(__clang__)) ||                              \
    defined(__MVS__)
#undef CRT_HAS_128BIT
#endif
```

> * The `.i` file claims to be from `compiler-rt/lib/builtins/divthc3.c`, which files doesn't exist upstream.  The body of the file isn't anything like `divtc3.c` either.

Sorry I included wrong file,  let me try again,  
[divtc3.c.txt](https://github.com/user-attachments/files/16627070/divtc3.c.txt)


> All the functions use (`__compiler_rt_logbtf` and several more) are guarded by `CRT_HAS_TF_MODE`. That works on SPARC, but won't for you since it isn't defined in your `.i` file. Are you maybe using modified sources?

Unfortunately, we don't have CRT_HAS_TF_MODE defined. I will check how come calling `__compiler_rt_logbtf1` is fine for us.

**Comment by zibi2** [comment]
His is the downstream code (not up streamed yet) guarded with `CRT_LDBL_128BIT`
```
# 488 "/plex/zibi/llvm/Woz/llvm-project/compiler-rt/lib/builtins/fp_lib.h"
static __inline tf_float __compiler_rt_logbtf(tf_float x) {
  return __builtin_logbl((x));
}
```
Maybe the guard for the `__divtc3` and `__multc3` should be:

`#if defined(CRT_HAS_F128) && (defined(CRT_HAS_TF_MODE) || defined(CRT_LDBL_128BIT))`

**Comment by arichardson** [comment]
If you can just call the `long double` libc versions, then it does sound like using 
`#if defined(CRT_HAS_TF_MODE) || defined(CRT_LDBL_128BIT)` as the guard would work.

This would be equivalent to `#if defined(CRT_HAS_F128) && (defined(CRT_HAS_128BIT) || defined(CRT_LDBL_128BIT))` but is IMO easier to understand.

**Comment by zibi2** [comment]
> If you can just call the `long double` libc versions, then it does sound like using `#if defined(CRT_HAS_TF_MODE) || defined(CRT_LDBL_128BIT)` as the guard would work.
> 
> This would be equivalent to `#if defined(CRT_HAS_F128) && (defined(CRT_HAS_128BIT) || defined(CRT_LDBL_128BIT))` but is IMO easier to understand.

We are fine with either one.

**Comment by rorth** [comment]
Neither of those works for 32-bit Solaris/sparc with gcc 14: the declarations of `__compiler_rt_logbtf` and friends in `fp_lib.h` all depend on `CRT_HAS_TF_MODE` being defined.  If it's not (as in the sparc case), we're again breaking what my original patch fixed.

The more I see this, the more I wonder if it's wise to make changes here that no in-tree target needs, arguing about a WIP z/OS port that may or may not need changes.  Maybe it's better to keep things as is for now and revisit this when that port is submitted for real and can be reviewed/tested as is, rather than arguing about unknown downstream code.  That would mean making whatever change is necessary to fix your issue locally for now.

**Comment by zibi2** [comment]
I would rather deal with this issue now when it's still fresh. Can you provide more details 

I'm confused, If `__compiler_rt_logbtf` depends on `CRT_HAS_TF_MODE` the following guard should work:
`#if defined(CRT_HAS_TF_MODE) || defined(CRT_LDBL_128BIT)`
Do you mean in 32-bit Solaris/sparc with gcc `CRT_HAS_TF_MODE` is not defined but `CRT_LDBL_128BIT` is and therefor the call to `__compiler_rt_logbtf` is unresolved? Is there any other macro we can use to distinguish this case? If you don't mind can you provide -dM -E output for 32-bit Solaris/sparc with gcc and the 64-bit one which I assume works?

**Comment by perry-ca** [comment]
The z/OS port is not a WIP.  It's real and we are in the process of upstreaming everything.  These breakages impact us and delay us being able to upstream everything.  Alex and I worked on getting things working.  I'd like to keep it working.

**Comment by perry-ca** [comment]
@rorth Can you try using the change in https://github.com/llvm/llvm-project/pull/82789 with your build.  I think that will solve your problem.  Some how that pr got closed instead of being merged.
